src/Java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/processing/GregtechMetaTileEntity_SpargeTower.java

```diff
@@ -0,0 +1,443 @@
+		.addOtherStructurePart("Sparge Tower Exterior Casing", "45 (minimum)")
+		.addEnergyHatch("Any casing", 1, 2)
+		.addMaintenanceHatch("Any casing", 1, 2, 3)
+		.addInputHatch("2x Input Hatches (Any bottom layer casing)", 1)
+		.addOutputHatch("6x Output Hatches (At least one per layer except bottom layer)", 2, 3)
+		.toolTipFinisher(CORE.GT_Tooltip_Builder);
+		return tt;
+	}
+
+	@Override
+	public ITexture[] getTexture(IGregTechTileEntity aBaseMetaTileEntity, byte aSide, byte aFacing, byte aColorIndex, boolean aActive, boolean aRedstone) {
+		if (aSide == aFacing) {
+			if (aActive)
+				return new ITexture[]{
+						BlockIcons.getCasingTextureForId(getCasingIndex()),
+						TextureFactory.builder().addIcon(TexturesGtBlock.Overlay_Machine_Controller_Advanced_Active).extFacing().build()};
+			return new ITexture[]{
+					BlockIcons.getCasingTextureForId(getCasingIndex()),
+					TextureFactory.builder().addIcon(TexturesGtBlock.Overlay_Machine_Controller_Advanced).extFacing().build()};
+		}
+		return new ITexture[]{Textures.BlockIcons.getCasingTextureForId(getCasingIndex())};
+	}
+
+	@Override
+	public Object getClientGUI(int aID, InventoryPlayer aPlayerInventory, IGregTechTileEntity aBaseMetaTileEntity) {
+		return new GT_GUIContainer_MultiMachine(aPlayerInventory, aBaseMetaTileEntity, getLocalName(), "DistillationTower.png");
+	}
+
+	@Override
+	public GT_Recipe.GT_Recipe_Map getRecipeMap() {
+		GTPP_Recipe_Map.sSpargeTowerRecipes.mRecipeList.clear();
+		if (GTPP_Recipe_Map.sSpargeTowerRecipes.mRecipeList.isEmpty()) {
+			generateRecipes();
+		}
+		return GTPP_Recipe_Map.sSpargeTowerRecipes;
+	}
+
+	private static boolean generateRecipes() {
+		for (GasSpargingRecipe aRecipe : GasSpargingRecipeMap.mRecipes) {
+			GTPP_Recipe newRecipe = new GTPP_Recipe(
+					false, 
+					new ItemStack[] {},
+					new ItemStack[] {},
+					null,
+					null,
+					aRecipe.mFluidInputs.clone(),
+					new FluidStack[] {},
+					aRecipe.mDuration, 
+					aRecipe.mEUt, 
+					0);
+			GTPP_Recipe_Map.sSpargeTowerRecipes.add(newRecipe);
+		}
+		if (GTPP_Recipe_Map.sSpargeTowerRecipes.mRecipeList.isEmpty()) {
+			return false;
+		}    	
+		return true;
+	}
+
+	@Override
+	public boolean isCorrectMachinePart(ItemStack aStack) {
+		return true;
+	}
+
+	@Override
+	public boolean checkRecipe(ItemStack aStack) {
+		ArrayList<FluidStack> tFluidList = getStoredFluids();
+		long tVoltage = getMaxInputVoltage();
+		byte tTier = (byte) Math.max(0, GT_Utility.getTier(tVoltage));
+		FluidStack[] tFluids = tFluidList.toArray(new FluidStack[0]);
+		if (tFluids.length > 0) {
+			Logger.INFO("Found "+tFluids.length+" input fluids. Searching "+GTPP_Recipe_Map.sSpargeTowerRecipes.mRecipeList.size()+" recipes.");
+			GT_Recipe tRecipe = getRecipeMap().findRecipe(getBaseMetaTileEntity(), false, gregtech.api.enums.GT_Values.V[tTier], tFluids);
+			if (tRecipe != null) {
+				Logger.INFO("Found recipe!");
+				try {
+					String[] aRecipeInfo = RecipeUtils.getRecipeInfo(tRecipe);
```

Comment:
Preferably not go over this array or generate it if debug logging is turned off.